lib/Parser/Parser.cpp

```diff
@@ -847,7 +868,23 @@ bool Parser::parseLine(std::string &ErrStr) {
       Block *B = 0;
 
       if (IK == Inst::Var) {
-        Inst *I = IC.createVar(InstWidth, InstName);
+        llvm::APInt Zero(InstWidth, 0, false), One(InstWidth, 0, false),
+                    ConstOne(InstWidth, 1, false);
+        if (CurTok.K == Token::KnownBits) {
+          if (InstWidth != CurTok.PatternString.length()) {
+            ErrStr = makeErrStr(TP, "knownbits pattern must be of same length as var width");
+            return false;
+          }
+          for (unsigned i=0; i<InstWidth; ++i) {
+            if (CurTok.PatternString[i] == '0')
+              Zero += (ConstOne.shl(CurTok.PatternString.length()-1-i));
+            else if (CurTok.PatternString[i] == '1')
+              One += (ConstOne.shl(CurTok.PatternString.length()-1-i));
```

Comment:
Yes, not required. I really need to be careful in my usage of parens and
braces. Sorry about that.

On Sun, Jul 19, 2015 at 11:23 PM, John Regehr <notifications@github.com>
wrote:

> In lib/Parser/Parser.cpp
> <https://github.com/google/souper/pull/195#discussion_r34968958>:
>
> > @@ -847,7 +868,23 @@ bool Parser::parseLine(std::string &ErrStr) {
> >        Block *B = 0;
> >
> >        if (IK == Inst::Var) {
> > -        Inst *I = IC.createVar(InstWidth, InstName);
> > +        llvm::APInt Zero(InstWidth, 0, false), One(InstWidth, 0, false),
> > +                    ConstOne(InstWidth, 1, false);
> > +        if (CurTok.K == Token::KnownBits) {
> > +          if (InstWidth != CurTok.PatternString.length()) {
> > +            ErrStr = makeErrStr(TP, "knownbits pattern must be of same length as var width");
> > +            return false;
> > +          }
> > +          for (unsigned i=0; i<InstWidth; ++i) {
> > +            if (CurTok.PatternString[i] == '0')
> > +              Zero += (ConstOne.shl(CurTok.PatternString.length()-1-i));
> > +            else if (CurTok.PatternString[i] == '1')
> > +              One += (ConstOne.shl(CurTok.PatternString.length()-1-i));
>
> I'm not sure the parens around the right-hand argument to += are helpful.
> Can you explain the purpose?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/souper/pull/195/files#r34968958>.
>



-- 
Thanks and Regards,
Jubi Taneja